### PR TITLE
{DO NOT MERGE} Feature/nopass - catch and print all exceptions

### PR DESCRIPTION
**Pull request by h4ck3rm1k3**
This is an experimental branch. Thanks for letting me use the jenkins.

I find it useful to know about problems. Salt squelches many of them for the developer of modules.

This pull request is not ready to merge et, but I wanted to open it up for discussion.
you will find a script checkexcept.py that looks for exceptions with no exception name, and then a runsed.sh that is generated to insert an exception handler that prints the error. 

You will find many patched files, more to come as I get more feedback and checks results from jenkins.

### CODE REVIEW

**Comment by cachedout** [comment]
Hi @h4ck3rm1k3 

Thanks for sending this in.

I think I see your concern here -- that a potentially harmful error might be ignored because exceptions may be silently ignored. Completely understandable. :]

However, a great number of exceptions in Python (and other programming languages) are raised as more of a signaling mechanism to callers than as actual errors. ``StopIteration`` is a classic example of this pattern and there are many others.

Consequently, simply assuming that a raised exception is an error which should be logged in the error log is not an approach that I believe is viable. In fact, doing so in an automated fashion may break Salt or hamper performance in Salt in some fairly substantial ways.

A large number of the cases that your script has identified have already been carefully considered by the original authors and undergone extensive code review to ensure that potentially useful information is not being lost. That said, mistakes happen and I absolutely believe there remain cases wherein additional logging may be helpful.

So, we're extremely hesitant to merge a pull-request that uses an automated approach to changing exception handling simply because so many of these cases have already been considered in the past.

We would be more than happy, however, to work with you on individual cases wherein additional logging might prove beneficial without breaking existing functionality or additional significant overhead.

**Comment by h4ck3rm1k3** [comment]
Alright, well iterators are a good example and they can be just filtered out using my tool. more to come.

**Comment by h4ck3rm1k3** [comment]
closing because this is going to take a while and be disruptive.

**Comment by h4ck3rm1k3** [comment]
I am reopening this because many of my tests are passing. I would like to see what is going on.

**Comment by UtahDave** [comment]
Hey, @h4ck3rm1k3, we're always VERY careful about large pull requests like this that modify hundreds of files. In our experience, making this many changes, especially automatically, is fraught with danger and unintended consequences. We have found that specific commits to fix a specific problem in the code are much more effective and less likely to cause problems than broad changes.

As we've already seen there are often reasons for the state of current code that is only clear when looking at the context of the lines in question. I'll defer to @cachedout and @terminalmage on these changes, but please be patient as we review this.  Thanks!

**Comment by thatch45** [comment]
@h4ck3rm1k3 I have to close this our from just a safety perspective, if we merge this it would cause us no end of problems, so just keeping it open strikes we as dangerous.
I understand where you are coming from on this, in that you believe that all exceptions are errors, but the reality is that the vast majority of exceptions in salt are there to manage the flow of the application (which is a perfectly reasonable use for exceptions) and managing how the plugin systems work(another reasonable case). If we were to follow this model we would also see a major hit to Salt performance.

I know that there are a number of cases where we should log the exceptions where we do currently not, but those cases are far smaller.

In a nutshell, please do not open PRs which have been automated on a grand scale like this in general, they are dangerous and always lead to breakage. I respect what you are trying to do, but your assumption is not a global case.
Thank you for your work, but this approach will lead us into many problems and cannot be sustained.

**Comment by h4ck3rm1k3** [comment]
Hello,
It is fine to close it, I never expected this to be merged, but wanted to use jenkins to run tests.
Here are my thoughts on this topic.
1. There changes are not for running in production, but for debugging problems in developing salt applications.
2. I have used my new patches to here and to jinja2 (https://github.com/electionr/jinja2/commits/passexception) to find problems with salt templates that stumped even the best of people who are totally familiar with salt. We found code that would never execute in templates that is not properly reported by salt. 
3. we need a strict/debug mode for salt that will report any and all errors.
4. we need a more declarative mode for files to declare what they are (states, modules etc) how they are used, what they depend on, what variables they use etc. This branch is going to turn into something that will find all the variables needed, all the modules needed for a bit of salt code and allow you to document it and create unit tests.
you can see my latest stab at unit tests here : https://github.com/h4ck3rm1k3/tomcat-formula/tree/master/tests I have created mock salt objects that allow for pillar and template evaluation and debugging. This coupled with magic mock will allow for documentation of interfaces and put an end to the function pointer model that I see here. I don't want to have a __salt__ object, but just import statements for the things I use and have check routines, best would be a static salt lint function.
So I am going to continue on this path out of my needs at work where we are attempting to use salt, because I don't see any way to use salt effectively with no error reporting.
